application/helpers/invoice_helper.php

```diff
@@ -128,4 +129,113 @@ function invoice_qrcode($invoice_id) {
+{
+    $CI = &get_instance();
+
+    $invoiceDateCreated = new DateTime(date_from_mysql($invoice_date_created, true));
+    
+    if ( 'system' == $client_language ) $dateLanguageLocale = get_setting('default_country');
+    else $dateLanguageLocale = invoice_get_locale_by_displayname($client_language);
+
+    // initialize based date
+    $printDate = clone($invoiceDateCreated);
+    // get the tags
+    $tags = explode('{{{', $item_description);
+    foreach ($tags as $tag) {
+        // find tags end
+        $rawTag = stristr($tag, '}}}', true);
+        // nothing to do here
+        if (empty($rawTag)) continue;
+
+        // here we do D(ate), M(onth and year) and Y(ear) nothing else
+        $request = strtoupper(substr($rawTag,0,1));  
+        // take only first letter, ignore if not within our service
+        if (strpos('DMY', $request) === false) continue;
+
+        // reconstruct original request pattern
+        $replaceThis='{{{'.$rawTag.'}}}';
+
+        // needs to reset if a new/second relative date occurs
+        // within same item description
+        try {
+            // calculate additions/substractions
+            if ($pos = strpos($rawTag, '+')) {
+                $num = substr($rawTag,$pos+1);
+                // refresh date to calculate with
+                $printDate = clone($invoice_date_created);
+                $printDate->add(new DateInterval( 'P' . $num . $request ));
+            }
+            elseif ($pos = strpos($rawTag, '-')) {
+                $num = substr($rawTag,$pos+1);
+                // refresh date to calculate with
+                $printDate = clone($invoice_date_created);
+                $printDate->sub(new DateInterval( 'P' . $num . $request ));
+            }
+            
+            // prepare replacement format string
+            if ('D' == $request) {
+                // ignore locale and create sql Y-m-d format date
+                // and use ip's function to create a visible date
+                $withReplacement = date_from_mysql(date_format($printDate, 'Y-m-d'));
+            }
+            else {
+                $dateFormat = new IntlDateFormatter(
+                                    $dateLanguageLocale,
+                                    IntlDateFormatter::SHORT,
+                                    IntlDateFormatter::SHORT
+                                    );
+                if ('M' == $request) $pattern = 'MMM yyyy'; // short month year
+                elseif ('Y' == $request) $pattern = 'yyyy';     // year only
+                $dateFormat->setPattern($pattern);
+                $withReplacement = datefmt_format($dateFormat, $printDate);
+            }
+
+            // replace within item description
+            $item_description = str_replace($replaceThis, $withReplacement, $item_description);
+
+        } catch (Exception $e) {
+            echo 'ERROR Message: ' .$e->getMessage();
+        }
+                
+    return $item_description;
+}
```

Comment:
There was a `}` missing.
I've fixed the code and made a PR on your branch on your repository.

I've also fixed the formatting.

If you could:

- [ ] double-check that function `invoice_replace_date_tags`
- [ ] see if my PR solves that problem

Then you can merge the PR and then tag me again. I'll make sure it ends up in InvoicePlane

Reply:
Iv'e done fixes and these are productive in my 1.5 version now.
There are samples in the templates for usage and a debugShowVars.php to help for own templates.
A README within templates explains usage of the helpers.

Reply:
Do you see a chance for short communication on phone? (might be native tongue)

Reply:
Not today, maybe this weekend